### ISSUE TITLE
collect_epoch_for_metrics — stream step outputs, auto-dedup, optional trim/shard

### ISSUE THREAD

**Issue by ishachat** [huggingface/accelerate#3749]
## Summary / problem
With even_batches=True (default), Accelerate keeps ranks aligned by duplicating a few samples at the tail. The docs recommend accelerator.gather_for_metrics(...) to gather per step and drop those duplicates. In practice, many users want an epoch-level result equal to len(dataset) without writing the same gather-append-concat boilerplate (and occasional final slice).

## Proposal
Add a helper that streams per-step outputs, uses gather_for_metrics internally, and returns an epoch-level tensor (exactly expected_len if provided). Optional: a callback to stream shards to disk instead of keeping everything in RAM.

##utils
collect_epoch_for_metrics(
    accelerator: Accelerator,
    step_iter: Iterable,
    step_fn: Callable[[Any], torch.Tensor],   # returns [b, ...] tensor per step
    *,
    cat_dim: int = 0,
    expected_len: Optional[int] = None,
    stream_to: Optional[Callable[[torch.Tensor, int], None]] = None,
) -> Optional[torch.Tensor]

##convenience method
Accelerator.collect_epoch_for_metrics(...)

## Semantics

Calls gather_for_metrics() each step → dedups tail automatically.
On main process: accumulates CPU chunks, torch.cat(..., dim=cat_dim) at the end.
If expected_len is set (or user passes len(dataset)), slice once as a final guard.
If stream_to is provided, call it with each gathered chunk and return a small summary (e.g., {"steps": k, "count": n}) instead of a giant tensor.

## Why useful

Collapses a very common eval/inference idiom into 1 call.
Reduces foot-guns (forgetting dedup/slice; holding everything on GPU).
Keeps low-level primitives (gather, gather_for_metrics) unchanged.

## Backwards compatibility
Purely additive. No behavior changes to existing APIs.
Test plan (CPU & single-process)
Concats multiple step outputs and returns sum(b_i) rows.
Honors expected_len by slicing.
stream_to gets called with correct step indices and shapes.
